{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "https://rgcgithub.github.io/regenie/options/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              |==================================|\n",
      "              |           REGENIE v1.0.4.1       |\n",
      "              |==================================|\n",
      "\n",
      "Copyright (c) 2020 Joelle Mbatchou and Jonathan Marchini.\n",
      "Distributed under the MIT License.\n",
      "\n",
      "Options:\n",
      " --help                            print list of available options\n",
      " --helpFull                        print list of all available options\n",
      " --step INT                        specify if fitting null model (=1) or association testing (=2)\n",
      " --bed PREFIX                      prefix to PLINK .bed/.bim/.fam files\n",
      " --pgen PREFIX                     prefix to PLINK2 .pgen/.pvar/.psam files\n",
      " --bgen FILE                       BGEN file\n",
      " --sample FILE                     sample file corresponding to BGEN file\n",
      " --keep FILE                       file listing samples to retain in the analysis\n",
      "                                   (no header; starts with FID IID)\n",
      " --remove FILE                     file listing samples to remove from the analysis\n",
      "                                   (no header; starts with FID IID)\n",
      " --extract FILE                    file with IDs of variants to retain in the analysis\n",
      " --exclude FILE                    file with IDs of variants to remove from the analysis\n",
      " --p FILE                          phenotype file (header required starting with FID IID)\n",
      " --phenoCol STRING                 phenotype name in header (use for each phenotype to keep)\n",
      " --phenoColList STRING             Comma separated list of phenotype names to keep\n",
      " --c FILE                          covariate file (header required starting with FID IID)\n",
      " --covarCol STRING                 covariate name in header (use for each covariate to keep)\n",
      " --covarColList STRING             Comma separated list of covariate names to keep\n",
      " --b INT                           size of genotype blocks\n",
      " --cv INT (=5)                     number of cross validation (CV) folds\n",
      " --l0 INT (=5)                     number of ridge parameters to use when fitting models\n",
      "                                   within blocks (evenly spaced in (0,1))\n",
      " --l1 INT (=5)                     number of ridge parameters to use when fitting model\n",
      "                                   across blocks (evenly spaced in (0,1))\n",
      " --loocv                           use leave-one out cross validation (LOOCV)\n",
      " --bt                              analyze phenotypes as binary (default is quantitative)\n",
      " --lowmem PREFIX                   reduce memory usage by writing level 0 predictions\n",
      "                                   to temporary files\n",
      " --threads INT (=ALL)              number of threads\n",
      " --strict                          remove all samples with missingness at any of the traits\n",
      " --o PREFIX                        prefix for output files\n",
      " --pred FILE                       file containing the list of files with predictions from step 1\n",
      " --ignore-pred                     skip reading predictions from step 1 (equivalent to\n",
      "                                   linear/logistic regression with only covariates)\n",
      " --force-impute                    keep and impute missing observations when in step 2 (default is\n",
      "                                   to drop missing for each trait)\n",
      " --minMAC INT (=5)                 minimum minor allele count (MAC) for tested variants\n",
      " --split                           split asssociation results into separate files for each trait\n",
      " --firth FLOAT (=0.05)             use Firth correction for p-values less than threshold\n",
      " --approx                          use approximation to Firth correction for computational speedup\n",
      " --spa FLOAT (=0.05)               use Saddlepoint approximation (SPA) for p-values less\n",
      "                                   than threshold\n",
      " --chr INT                         specify chromosome to test in step 2 (use for each chromosome)\n",
      " --chrList STRING                  Comma separated list of chromosomes to test in step 2\n",
      " --test STRING                     specify to use dominant or recessive test\n",
      "\n",
      "For more information, visit the website: https://rgcgithub.github.io/regenie/\n"
     ]
    }
   ],
   "source": [
    "!regenie --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start time: Sat Jul 11 18:46:29 2020\n",
      "\n",
      "              |==================================|\n",
      "              |           REGENIE v1.0.4.1       |\n",
      "              |==================================|\n",
      "\n",
      "Copyright (c) 2020 Joelle Mbatchou and Jonathan Marchini.\n",
      "Distributed under the MIT License.\n",
      "\n",
      "Log of output saved in file : fit_bin_out.log\n",
      "\n",
      "Command line arguments:\n",
      "  --step 1 \n",
      "  --bed /home/jovyan/work/repos/regenie/example/example \n",
      "  --exclude /home/jovyan/work/repos/regenie/example/snplist_rm.txt \n",
      "  --c /home/jovyan/work/repos/regenie/example/covariates.txt \n",
      "  --p /home/jovyan/work/repos/regenie/example/phenotype_bin.txt \n",
      "  --remove /home/jovyan/work/repos/regenie/example/fid_iid_to_remove.txt \n",
      "  --b 100 \n",
      "  --bt \n",
      "  --lowmem \n",
      "  --o fit_bin_out \n",
      "\n",
      "Fitting null model\n",
      " * bim              : [/home/jovyan/work/repos/regenie/example/example.bim] n_snps = 1000\n",
      "   -removing variants specified in [/home/jovyan/work/repos/regenie/example/snplist_rm.txt]\n",
      "   -number of variants remaining in the analysis = 994\n",
      " * fam              : [/home/jovyan/work/repos/regenie/example/example.fam] n_samples = 500\n",
      " * bed              : [/home/jovyan/work/repos/regenie/example/example.bed]\n",
      "   -removing individuals specified in [/home/jovyan/work/repos/regenie/example/fid_iid_to_remove.txt]\n",
      "     +number of genotyped individuals to exclude from the analysis = 6\n",
      " * phenotypes       : [/home/jovyan/work/repos/regenie/example/phenotype_bin.txt] n_pheno = 2\n",
      "   -number of phenotyped individuals = 494\n",
      " * covariates       : [/home/jovyan/work/repos/regenie/example/covariates.txt] n_cov = 3\n",
      "   -number of individuals with covariate data = 494\n",
      " * number of individuals used in analysis = 494\n",
      "   -fitting null logistic regression on binary phenotypes...done (0ms) \n",
      "   -residualizing and scaling phenotypes...done (0ms) \n",
      "   -WARNING: Sample size is less than 5,000 so using LOOCV instead of 5-fold CV.\n",
      " * # threads        : [16]\n",
      " * block size       : [100]\n",
      " * # blocks         : [10]\n",
      " * # CV folds       : [494]\n",
      " * ridge data_l0    : [5 : 0.01 0.25 0.5 0.75 0.99 ]\n",
      " * ridge data_l1    : [5 : 0.01 0.25 0.5 0.75 0.99 ]\n",
      " * approximate memory usage : 955MB\n",
      " * writing level 0 predictions to disk\n",
      "   -temporary files will have prefix [fit_bin_out_l0_Y]\n",
      "   -approximate disk space needed : 1MB\n",
      " * setting memory...done\n",
      "\n",
      "Chromosome 1\n",
      " block [1] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (1ms) \n",
      "   -calc working matrices...done (8ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [2] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (9ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [3] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (19ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [4] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (13ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [5] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (18ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [6] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (13ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [7] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (13ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [8] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (13ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [9] : 100 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (13ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      " block [10] : 94 snps  (0ms) \n",
      "   -residualizing and scaling genotypes...done (0ms) \n",
      "   -calc working matrices...done (25ms) \n",
      "   -calc level 0 ridge...done (1ms) \n",
      "\n",
      " Level 1 ridge with logistic regression...\n",
      "   -on phenotype 1 (Y1)...done (9ms) \n",
      "   -on phenotype 2 (Y2)...done (6ms) \n",
      "\n",
      "Output\n",
      "------\n",
      "phenotype 1 (Y1) : \n",
      "  0.01  : Rsq = 0.368362, MSE = 0.116198, -logLik/N = 0.356728<- min value\n",
      "  0.25  : Rsq = 0.361251, MSE = 0.117645, -logLik/N = 0.362252\n",
      "  0.5   : Rsq = 0.360409, MSE = 0.11784, -logLik/N = 0.363211\n",
      "  0.75  : Rsq = 0.363963, MSE = 0.117171, -logLik/N = 0.361939\n",
      "  0.99  : Rsq = 0.385039, MSE = 0.113323, -logLik/N = 0.357031\n",
      "  * making predictions...writing LOCO predictions...done (6ms) \n",
      "\n",
      "phenotype 2 (Y2) : \n",
      "  0.01  : Rsq = 0.441675, MSE = 0.0991279, -logLik/N = 0.320683\n",
      "  0.25  : Rsq = 0.436439, MSE = 0.100057, -logLik/N = 0.322382\n",
      "  0.5   : Rsq = 0.434385, MSE = 0.10044, -logLik/N = 0.323374\n",
      "  0.75  : Rsq = 0.437287, MSE = 0.0999308, -logLik/N = 0.322161\n",
      "  0.99  : Rsq = 0.456629, MSE = 0.0966395, -logLik/N = 0.315832<- min value\n",
      "  * making predictions...writing LOCO predictions...done (6ms) \n",
      "\n",
      "List of blup files written to: [fit_bin_out_pred.list] \n",
      "\n",
      "Elapsed time : 0.221319s\n",
      "End time: Sat Jul 11 18:46:29 2020\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "EXDIR=$WORK_DIR/repos/regenie/example\n",
    "regenie \\\n",
    "  --step 1 \\\n",
    "  --bed $EXDIR/example \\\n",
    "  --exclude $EXDIR/snplist_rm.txt \\\n",
    "  --c $EXDIR/covariates.txt \\\n",
    "  --p $EXDIR/phenotype_bin.txt \\\n",
    "  --remove $EXDIR/fid_iid_to_remove.txt \\\n",
    "  --b 100 \\\n",
    "  --bt --lowmem \\\n",
    "  --o fit_bin_out"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
